### PR TITLE
 [BEAM-3207] Create a standard location to enumerate and document URNs.

### PR DESCRIPTION
**Pull request by robertwb**
URNs are listed in a markdown file in the pipeline definitions module.
This file is used to auto-generate URN constants for the Python SDK and
validate URN constants in the Java SDK (though eventually it'd be good
to auto-generate them in this case as well).

The format of these common URNs has been normalized to

    org.apache.beam:type:name:vN[.M]

SDK-specific URNs are left as they are.

Further fleshing out the definitions and specifications of all these URNS,
as well as making sure they are used ubiquitiously, can be deferred to
a later PR now that there is a central location to work from.

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by robertwb** [comment]
Jenkins retest this please. (ERROR: beam8 is offline)

**Comment by robertwb** [comment]
Jenkins retest this please

**Comment by robertwb** [comment]
Jenkins retest this please

File: model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+###  org.apache.beam:coder:global_window:v1
+
+###  org.apache.beam:coder:interval_window:v1
+
+###  org.apache.beam:coder:windowed_value:v1
+
+
+## Side input access
+
+### org.apache.beam:sideinput:iterable:v1
```
Comment (kennknowles) [inline model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md]: nit: in other places words are underscore-separated

File: model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+-->
+
+# Apache Beam URNs
+
+This file serves as a central place to enumerate and document the various
```
Comment (kennknowles) [inline model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md]: Maybe YAML would allow easier parsing and machine-readable association of metadata and commentary with the URNs? This file already suggests a `comment` and `payload` field for general description and a machine-readable spec for the payload.

File: model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+URNs used in the Beam portability APIs.
+
+
+## Core Transforms
+
+### org.apache.beam:transform:pardo:v1
```
Comment (kennknowles) [inline model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md]: Technically, a URN has the form `urn:` _<namespace>_ `:` _<namespace-specific section>_

So, we drop the `urn` since it is redundant from context and I'm not sure we really need `org.apache.beam` as the namespace. It isn't an "authority" section as in other URIs. We can just use `beam` as we already have in most of the constants. The actual process is just that you get it registered with IANA [here](https://www.iana.org/assignments/urn-namespaces/urn-namespaces.xhtml) but I don't know that we need to. TBH even "beam" is probably redundant when we are putting them into URN fields in a Beam proto.

File: .gitignore [hunk 4]
```diff
@@ -34,6 +34,7 @@ sdks/python/LICENSE
 sdks/python/NOTICE
 sdks/python/README.md
 sdks/python/apache_beam/portability/api/*pb2*.*
+sdks/python/apache_beam/portability/common_urns.py
```
Comment (kennknowles) [inline .gitignore]: Is it necessary to have a generated file? Can't you just reflectively and lazily generate the module/class/constants? (whichever is easiest)

File: model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+###  org.apache.beam:coder:global_window:v1
+
+###  org.apache.beam:coder:interval_window:v1
+
+###  org.apache.beam:coder:windowed_value:v1
+
+
+## Side input access
+
+### org.apache.beam:side_input:iterable:v1
```
Comment (lukecwik) [inline model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md]: This one shouldn't exist/be removed.

File: model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md (hunk 3 above)
Comment (lukecwik) [reply model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md]: Should we follow semantic versioning for the version numbers similar to how we are releasing the SDK versions (e.g. org.apache.beam:transform:pardo:1.0.0)?

Comment (robertwb) [reply model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md]: Currently we're just requiring exact matches. I'm not sure what a "bugfix" version means for URNs, I suppose minor upgrades could mean adding formerly optional fields to proto-like payloads, but that could be checked for directly as well.

File: model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md (hunk 5 above)
Comment (robertwb) [reply model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md]: I was looking at what it would take to remove this from the Python SDK, and given that the existing batch code doesn't support indexable side inputs I think we should offer this option as well. (This seems a better option than generating different graphs when trying to run portably.)

File: model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md (hunk 3 above)
Comment (kennknowles) [reply model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md]: It is pretty common for protocols to use the minor version, but I don't know of any that use the bugfix version.

There's a polarity issue in the higher-order case that seems to also apply here - adding a field to an interface is backwards compatible for consumers, but incompatible for implementers of the interface. But adding an optional field is both backward and forward compatible so you don't need a minor version bump.

File: model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md (hunk 5 above)
Comment (lukecwik) [reply model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md]: How about we drop it for now and just use it inside the Python SDK till there is a need. I don't want this to make its way out of the Python SDK till there is a strong need for iterable side inputs over the Fn API.

**Comment by robertwb** [comment]
Updated per discussion. PTAL.

**Comment by robertwb** [comment]
Jenkins: retest this please. (Backing channel 'beam8' is disconnected)

**Comment by robertwb** [comment]
Jenkins: retest this please.

**Comment by robertwb** [comment]
Rebased again due to merge conflicts. It would be good to get this in.

**Comment by kennknowles** [comment]
Yea, many apologies for the delay. Review load is very high lately (a good thing).

**Comment by kennknowles** [comment]
It does look like the errors may be caused by things introduced since you wrote this. I'm jumping to that conclusion because I saw mention of URNs in the error messages...

**Comment by robertwb** [comment]
Jenkins: retest this please.

**Comment by robertwb** [comment]
I have resolved the issues with the Dataflow runner, opting to leave those URNs currently hard coded in the workers as they are in this PR (which is already big and painful enough as it is) and defer the worker dance to a future one.

PTAL

**Comment by kennknowles** [comment]
I'm wary about `setup.py` because "actually programming" in builds is one of the worst sources of unpredictable maintenance burdens, but let's just get this in and iterate because it is definitely useful immediately. The gradle failure is [BEAM-3605](https://issues.apache.org/jira/browse/BEAM-3605).

**Comment by robertwb** [comment]
Thanks!

File: sdks/python/apache_beam/transforms/core.py [hunk 6]
```diff
@@ -1435,18 +1429,6 @@ def expand(self, pcoll):
     self._check_pcollection(pcoll)
     return pvalue.PCollection(pcoll.pipeline)
 
-  def to_runner_api_parameter(self, context):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/transforms/core.py]: Why was this removed?  This change conflicts with the changes in https://github.com/apache/incubator-beam/pull/4529.
